MSMetaEnhancer/libs/utils/Monitor.py

```diff
@@ -0,0 +1,58 @@
+from threading import Thread, Event
+from urllib.parse import urlparse
+import time
+import requests
+
+
+class Monitor(Thread):
+    """
+    Class to periodically monitor status of used services.
+    """
+    def __init__(self, converters):
+        super(Monitor, self).__init__()
+        self.converters = converters
+        self.stop_request = Event()
+        self.first_check = Event()
+
+    @staticmethod
+    def get_base_url(converter):
+        """
+        Extract base URL from the converter.
+
+        :param converter: given service
+        :return: base URL
+        """
+        url = urlparse(list(converter.services.values())[0])
+        return url.scheme + "://" + url.netloc
+
+    @staticmethod
+    def check_service(url):
+        """
+        Send a GET request to given URL with 5 sec timeout.
+
+        :param url: given URL
+        :return: True if request is successful with status code 200
+        """
+        try:
+            result = requests.get(url, timeout=5)
+            return result.status_code == 200
+        except (requests.exceptions.ConnectionError, TimeoutError):
+            return False
+
+    def run(self):
+        """
+        Main loop of the Monitor thread.
+
+        Monitor checks if GET on base URL of every service give status code 200.
+        Such a service is considered available.
+        This is checked periodically to always have up-to-date information.
+        """
+        while not self.stop_request.isSet():
+            for converter in self.converters.values():
+                url = self.get_base_url(converter)
+                converter.is_available = self.check_service(url)
+            self.first_check.set()
+            time.sleep(10)
```

Comment:
This seems quite excessive. My reasoning follows:

- Imagine 100 jobs run at the same time on various Galaxies. How many "alive" requests are the services receiving from MSME?
- I am not convinced we have a good response to a service that goes down during the job execution anyways. I cannot see how we can currently resume it later or fall back on some other service.

My simplistic approach to the unreliability of services problem would be: "service is down = job will fail = please rerun it later".

If we have more data on how often and for how long the services stop responding, we could devise a smarter strategy of e.g. restarting this on user's behalf or pausing the whole process but at the moment I do not see a clear benefit of this implementation.

Reply:
The benefit of this implementation is performance-wise, trying to avoid excessive requests on a service which is for example time-outing. Instead of sending the request for every scpectra in the file and waiting, the particular job using this service will be skipped. 

The information about it could be passed to the user, probably using [log](https://github.com/RECETOX/galaxytools/issues/201).

Reply:
To be realistic, I think this tool will be used much less than the actual web interface, so pinging them every 10s while running a job is not going to create a serious overhead. I mean we could even make the monitor also ping or make a post with a specific token and then inform the service providers that if this is causing too much traffic, we can change it. For now I wouldn't worry about whether this will run with thousands of user's - this will very likely never be the case.